React/Base/RCTComponentEvent.m

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "RCTComponentEvent.h"
+
+#import "RCTAssert.h"
+
+@implementation RCTComponentEvent
+{
+  NSArray *_arguments;
+}
+
+@synthesize eventName = _eventName;
+@synthesize viewTag = _viewTag;
+
+- (instancetype)initWithName:(NSString *)name viewTag:(NSNumber *)viewTag body:(NSDictionary *)body
+{
+  if (self = [super init]) {
+    NSMutableDictionary *mutableBody = [NSMutableDictionary dictionaryWithDictionary:body];
```

Comment:
Wait, that does not look right. Why do we need put tag inside body? That was needed for old poorly designed `sendEventBlahBlahWithoutTagButWithBody:`, right?

Reply:
I'm pretty sure JS needs the tag in the body that's why we put it here. I could be wrong though I'd have to look at the RCTEventEmitter.receiveEvent code.

Reply:
Yea it is definitely used https://github.com/facebook/react/blob/7f78749ee3023608a614b2f595739bcd36acc128/src/renderers/native/ReactNativeEventEmitter.js#L112. I'd rather not touch this :)

Reply:
Yeah... But... I am trying understand that code... and I failed.
Looks like we are sending `tag` as `rootNodeID` (which is not even tag!) as a first argument... Why? 🤔

Reply:
Yea I'm not sure why we are sending the tag as both the first arg and in the event body as target. I just think we should keep it like this to avoid breaking react.

Reply:
First of all, I am perfectly okay with applying our/your change iteratively. We don't need to modify half of react event subsystem to remove just one deprecated method, we don't need to it in single diff. However, we have to have good reasoning behind our changes and good plan for future changes.
So, let's play with debugger and verify that evething works as we think it works at least at same not-so-deep levels of abstractions, so verify that `tag` really became `rootNodeID`. (I also will investigate this part.) Then we have to explicitly mention that controversy in code and mark that as temporary legacy leftovers. Land this. And continue to improve the architecture of RN.

I am very exited about this diff!

Reply:
@bvaughn Could you clarify why we need to pass both `rootNodeID` and `nativeEvent.target` here https://github.com/facebook/react/blob/7f78749ee3023608a614b2f595739bcd36acc128/src/renderers/native/ReactNativeEventEmitter.js#L126? AFAIK those 2 values are always the same.

Reply:
@shergin Let's see if someone from the react team can help here. I agree we should have a clear plan here. We also have to consider android if we make any change to the native / react bridging api.

Reply:
Hi! I'm only somewhat familiar with our event code. (I don't think anyone on the core team presently is deeply familiar with it, to be honest.)

The event params like `target` _seem_ to be optional given the [Flow typing](https://github.com/facebook/react/blob/master/src/renderers/native/ReactNativeEventEmitter.js#L103) and the fact that we [fallback to undefined values](https://github.com/facebook/react/blob/master/src/renderers/native/ReactNativeEventEmitter.js#L105).

If the 2 values _are_ always the same (for Android too) then it seems silly to always pass them both and I could update the `ReactNativeEventEmitter` to just create this wrapper object (or not, if we determine it's not actually being used for anything).

Reply:
FWIW it looks like the Java implementation doesn't necessarily pass a `target` but it does pass other attributes in some cases (eg [on content size change](https://github.com/facebook/react-native/blob/1e8f3b11027fe0a7514b4fc97d0798d3c64bc895/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/ContentSizeChangeEvent.java#L31-L36)). The Java interface also [defines the map as a `@Nullable` type](https://github.com/facebook/react-native/blob/1e8f3b11027fe0a7514b4fc97d0798d3c64bc895/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/RCTEventEmitter.java#L19) and in [at least some cases](https://github.com/facebook/react-native/blob/1e8f3b11027fe0a7514b4fc97d0798d3c64bc895/ReactAndroid/src/main/java/com/facebook/react/views/modal/RequestCloseEvent.java#L33) explicitly passes `null`.

Reply:
Thanks @bvaughn, I'm pretty sure we need to keep target in the event payload otherwise it would be a breaking change for apps that rely on this value. Maybe we could avoir passing the tag as the first argument if it's always the same as body.target. Needs more investigation.

@shergin What do you think about landing this as is since it will unblock a few other improvements I wanted to make?

Reply:
> I'm pretty sure we need to keep target in the event payload otherwise it would be a breaking change for apps that rely on this value. 

Sure! To be clear: I wasn't suggesting we remove `event.target` on the JS side, just that we could remove the duplicate parameter and (on the RN JavaScript side) just always set target equal to the tag number.